src/net/connection/connection_pool.rs

```diff
@@ -0,0 +1,95 @@
+use super::{ConnectionTimeout, VirtualConnection};
+use error::{NetworkError, Error,Result};
+
+use std::thread;
+use std::time::Duration;
+use std::sync::{Arc, Mutex, RwLock, RwLockWriteGuard};
+use std::net::{SocketAddr};
+use std::collections::HashMap;
+
+// Default timeout of a specific client
+const TIMEOUT_DEFAULT: ConnectionTimeout = 10;
+
+// Default time between checks of all clients for timeouts in seconds
+const TIMEOUT_POLL_INTERVAL: u64 = 1;
+
+/// This is an pool of virtual connections (connected) over UDP.
+pub struct ConnectionPool
+{
+    timeout: ConnectionTimeout,
+    connections: Arc<RwLock<HashMap<SocketAddr, Arc<RwLock<VirtualConnection>>>>>,
+    sleepy_time: Duration,
+    poll_interval: Duration,
+}
+
+impl ConnectionPool {
+    pub fn new() -> ConnectionPool
+    {
+        let sleepy_time = Duration::from_secs(TIMEOUT_DEFAULT);
+        let poll_interval = Duration::from_secs(TIMEOUT_POLL_INTERVAL);
+
+        ConnectionPool { timeout: TIMEOUT_DEFAULT, connections: Arc::new(RwLock::new(HashMap::new())), sleepy_time, poll_interval }
+    }
+
+    /// Set the timeout before an client will be seen as disconnected.
+    pub fn set_timeout(&mut self, timeout: ConnectionTimeout)
+    {
+        self.timeout = timeout;
+    }
+
+    /// Insert connection if it does not exists.
+    pub fn insert_if_not_exists(&mut self, addr: &SocketAddr) -> Result<Arc<RwLock<VirtualConnection>>> {
```

Comment:
Please rename this to `get_connection_or_insert()` and use https://github.com/amethyst/laminar/blob/master/src/net/socket_state.rs#L220 as an example to implement this with proper read and write lock usage.